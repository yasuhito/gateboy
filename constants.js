export const COLS = 10;
export const ROWS = 20;
const BLOCK_SIZE = 30;
export const NO_OF_HIGH_SCORES = 10;
export const HIGH_SCORES = 'highScores';
export const NAMES = ['H', 'X', 'Y', 'Z', 'S', 'T']
export const COLORS = ['cyan', 'blue', 'orange', 'green', 'purple', 'red'];

// const SHAPES = [
//   [['H']],
//   [['X']],
//   [['Y']],
//   [['Z']],
//   [['S']],
//   [['T']]
// ];

export const SHAPES = [
  [['I', 'I', 'I', 'I'], ['H', 'H', 'H', 'H'], ['I', 'I', 'I', 'I'], ['I', 'I', 'I', 'I']],
  [['X', 'I', 'I'], ['X', 'X', 'X'], ['I', 'I', 'I']],
  [['I', 'I', 'Y'], ['Y', 'Y', 'Y'], ['I', 'I', 'I']],
  [['Z', 'Z'], ['Z', 'Z']],
  [['I', 'S', 'S'], ['S', 'S', 'I'], ['I', 'I', 'I']],
  [['I', 'T', 'I'], ['T', 'T', 'T'], ['I', 'I', 'I']],
  // [[7, 7, 'I'], ['I', 7, 7], ['I', 'I', 'I']]
];

export const POINTS = {
  SINGLE: 100,
  DOUBLE: 300,
  TRIPLE: 500,
  TETRIS: 800,
  SOFT_DROP: 1,
  HARD_DROP: 2
}
Object.freeze(POINTS);

export const GATES_PER_LEVEL = 10;
export const LEVEL = {
  0: 800,
  1: 720,
  2: 630,
  3: 550,
  4: 470,
  5: 380,
  6: 300,
  7: 220,
  8: 130,
  9: 100,
  10: 80,
  11: 80,
  12: 80,
  13: 70,
  14: 70,
  15: 70,
  16: 50,
  17: 50,
  18: 50,
  19: 30,
  20: 30,
  // 29+ is 20ms
}
Object.freeze(LEVEL);

const canvas = document.getElementById('board');
export const ctx = canvas.getContext('2d');
const canvasNext = document.getElementById('next');
export const ctxNext = canvasNext.getContext('2d');

// Calculate size of canvas from constants.
ctx.canvas.width = COLS * BLOCK_SIZE;
ctx.canvas.height = ROWS * BLOCK_SIZE;

// Size canvas for four blocks.
ctxNext.canvas.width = 4 * BLOCK_SIZE;
ctxNext.canvas.height = 4 * BLOCK_SIZE;

// Scale blocks
ctx.scale(BLOCK_SIZE, BLOCK_SIZE);
ctxNext.scale(BLOCK_SIZE, BLOCK_SIZE);


export const KEY = {
  SPACE: 32,
  LEFT: 37,
  UP: 38,
  RIGHT: 39,
  DOWN: 40
}
Object.freeze(KEY);

export const moves = {
  [KEY.LEFT]:  (board) => ({ ...board.gate, x: board.gate.x - 1 }),
  [KEY.RIGHT]: (board) => ({ ...board.gate, x: board.gate.x + 1 }),
  [KEY.DOWN]:  (board) => ({ ...board.gate, y: board.gate.y + 1 }),
  [KEY.UP]:    (board) => board.rotate(board.gate),
  [KEY.SPACE]:  (board) => ({ ...board.gate, y: board.gate.y + 1 }),
};