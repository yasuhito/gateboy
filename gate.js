class Gate {
  constructor(ctx) {
    this.ctx = ctx;

    const typeId = this.randomizeTetrominoType(COLORS.length);
    this.name = NAMES[typeId];
    this.shape = SHAPES[typeId];
    this.color = COLORS[typeId];

    this.x = 0;
    this.y = 0;
  }

  draw() {
    this.ctx.fillStyle = this.color;
    this.shape.forEach((row, y) => {
      row.forEach((value, x) => {
        if (value > 0) {
          this.ctx.fillRect(this.x + x, this.y + y, 1, 1);
        }
      });
    });

    // draw gate nae
    this.ctx.fillStyle = '#000';
    this.ctx.font = '1px sans-serif';
    this.ctx.fillText(this.name, this.x, this.y + 1);
  }

  move(gate) {
    this.x = gate.x;
    this.y = gate.y;
    this.shape = gate.shape;
  }

  // TODO: implement TGM (The Grand Master series) randomizer
  randomizeTetrominoType(noOfTypes) {
    return Math.floor(Math.random() * noOfTypes);
  }
}
