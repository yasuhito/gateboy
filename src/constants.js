export const NO_OF_HIGH_SCORES = 10;
export const HIGH_SCORES = "highScores";

export const POINTS = {
  SINGLE: 100,
  DOUBLE: 300,
  TRIPLE: 500,
  TETRIS: 800,
  SOFT_DROP: 1,
  HARD_DROP: 2,
};
Object.freeze(POINTS);

export const GATES_PER_LEVEL = 10;
export const LEVEL = {
  0: 800,
  1: 720,
  2: 630,
  3: 550,
  4: 470,
  5: 380,
  6: 300,
  7: 220,
  8: 130,
  9: 100,
  10: 80,
  11: 80,
  12: 80,
  13: 70,
  14: 70,
  15: 70,
  16: 50,
  17: 50,
  18: 50,
  19: 30,
  20: 30,
  // 29+ is 20ms
};
Object.freeze(LEVEL);

export const KEY = {
  SPACE: 32,
  LEFT: 37,
  UP: 38,
  RIGHT: 39,
  DOWN: 40,
};
Object.freeze(KEY);

export const moves = {
  [KEY.LEFT]: (board) => ({ ...board.gate, x: board.gate.x - 1 }),
  [KEY.RIGHT]: (board) => ({ ...board.gate, x: board.gate.x + 1 }),
  [KEY.DOWN]: (board) => ({ ...board.gate, y: board.gate.y + 1 }),
  [KEY.UP]: (board) => board.rotate(board.gate),
  [KEY.SPACE]: (board) => ({ ...board.gate, y: board.gate.y + 1 }),
};
