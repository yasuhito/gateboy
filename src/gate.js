const NAMES = ['H', 'X', 'Y', 'Z', 'S', 'T']
const SHAPES = [
  [['I', 'I', 'I', 'I'], ['H', 'H', 'H', 'H'], ['I', 'I', 'I', 'I'], ['I', 'I', 'I', 'I']],
  [['X', 'I', 'I'], ['X', 'X', 'X'], ['I', 'I', 'I']],
  [['I', 'I', 'Y'], ['Y', 'Y', 'Y'], ['I', 'I', 'I']],
  [['Z', 'Z'], ['Z', 'Z']],
  [['I', 'S', 'S'], ['S', 'S', 'I'], ['I', 'I', 'I']],
  [['I', 'T', 'I'], ['T', 'T', 'T'], ['I', 'I', 'I']],
  // [[7, 7, 'I'], ['I', 7, 7], ['I', 'I', 'I']]
];
const COLORS = ['cyan', 'blue', 'orange', 'green', 'purple', 'red'];

export class Gate {
  static random() {
    const typeId = this.randomizeGateType();
    const gate = new Gate(this.NAMES[typeId])

    return gate
  }

  // TODO: implement TGM (The Grand Master series) randomizer
  static randomizeGateType() {
    const noOfTypes = Gate.NAMES.length;
    return Math.floor(Math.random() * noOfTypes);
  }

  static get NAMES() {
    return NAMES;
  }

  static get SHAPES() {
    return SHAPES;
  }

  static get COLORS() {
    return COLORS;
  }

  constructor(name) {
    const typeId = Gate.NAMES.indexOf(name)

    this.name = name
    this.shape = Gate.SHAPES[typeId];
    this.color = Gate.COLORS[typeId];
    this.x = 0;
    this.y = 0;
  }

  draw(ctx) {
    this.shape.forEach((row, y) => {
      row.forEach((value, x) => {
        if (value === 'I') return;

        // draw gate rectangle
        ctx.fillStyle = this.color;
        ctx.fillRect(this.x + x, this.y + y, 1, 1);

        // draw gate name
        ctx.fillStyle = '#000';
        ctx.font = '1px sans-serif';
        ctx.fillText(this.name, this.x + x, this.y + y + 1);
      });
    });
  }

  move(gate) {
    this.x = gate.x;
    this.y = gate.y;
    this.shape = gate.shape;
  }
}
