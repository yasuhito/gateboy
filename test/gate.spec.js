import {Gate} from '../src/gate.js';
import {getBoardCanvasContext} from '../src/constants.js';

describe('Gate', () => {
  beforeEach(function() {
    const board = document.createElement('canvas');
    board.id = 'board'
    document.body.appendChild(board);
  });

  describe('new', () => {
    it('should have name', function () {
      const gate = new Gate('H');

      expect(gate.name).toEqual('H');
    })

    it('should have shape', function () {
      const gate = new Gate('H');

      expect(gate.shape).toEqual([
        ['I', 'I', 'I', 'I'],
        ['H', 'H', 'H', 'H'],
        ['I', 'I', 'I', 'I'],
        ['I', 'I', 'I', 'I']
      ]);
    })

    it('should have color', function () {
      const gate = new Gate('H');

      expect(gate.color).toEqual('cyan');
    })

    it('should initialize x and y with 0', function () {
      const gate = new Gate('H');

      expect(gate.x).toBe(0)
      expect(gate.y).toBe(0)
    })
  })

  describe('draw', () => {
    it('should draw its shape without errors', function () {
      const gate = new Gate('H');
      const ctx = getBoardCanvasContext();

      gate.draw(ctx)
      const events = ctx.__getEvents();

      expect(events).toMatchSnapshot();
    })
  })
})
