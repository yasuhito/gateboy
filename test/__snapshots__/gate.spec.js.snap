// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gate draw should draw its shape without errors 1`] = `
Array [
  Object {
    "props": Object {
      "x": 30,
      "y": 30,
    },
    "transform": Array [
      30,
      0,
      0,
      30,
      0,
      0,
    ],
    "type": "scale",
  },
  Object {
    "props": Object {
      "value": "#00ffff",
    },
    "transform": Array [
      30,
      0,
      0,
      30,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "height": 1,
      "width": 1,
      "x": 0,
      "y": 1,
    },
    "transform": Array [
      30,
      0,
      0,
      30,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "value": "#000000",
    },
    "transform": Array [
      30,
      0,
      0,
      30,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "value": "1px sans-serif",
    },
    "transform": Array [
      30,
      0,
      0,
      30,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "H",
      "x": 0,
      "y": 2,
    },
    "transform": Array [
      30,
      0,
      0,
      30,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "value": "#00ffff",
    },
    "transform": Array [
      30,
      0,
      0,
      30,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "height": 1,
      "width": 1,
      "x": 1,
      "y": 1,
    },
    "transform": Array [
      30,
      0,
      0,
      30,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "value": "#000000",
    },
    "transform": Array [
      30,
      0,
      0,
      30,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "value": "1px sans-serif",
    },
    "transform": Array [
      30,
      0,
      0,
      30,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "H",
      "x": 1,
      "y": 2,
    },
    "transform": Array [
      30,
      0,
      0,
      30,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "value": "#00ffff",
    },
    "transform": Array [
      30,
      0,
      0,
      30,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "height": 1,
      "width": 1,
      "x": 2,
      "y": 1,
    },
    "transform": Array [
      30,
      0,
      0,
      30,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "value": "#000000",
    },
    "transform": Array [
      30,
      0,
      0,
      30,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "value": "1px sans-serif",
    },
    "transform": Array [
      30,
      0,
      0,
      30,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "H",
      "x": 2,
      "y": 2,
    },
    "transform": Array [
      30,
      0,
      0,
      30,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "value": "#00ffff",
    },
    "transform": Array [
      30,
      0,
      0,
      30,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "height": 1,
      "width": 1,
      "x": 3,
      "y": 1,
    },
    "transform": Array [
      30,
      0,
      0,
      30,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "value": "#000000",
    },
    "transform": Array [
      30,
      0,
      0,
      30,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "value": "1px sans-serif",
    },
    "transform": Array [
      30,
      0,
      0,
      30,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "H",
      "x": 3,
      "y": 2,
    },
    "transform": Array [
      30,
      0,
      0,
      30,
      0,
      0,
    ],
    "type": "fillText",
  },
]
`;
